### PR TITLE
Add various tweaks and improvements to CSV parsing

### PR DESCRIPTION
**Pull request by sburns**
* Redirect when no session data after bulk upload
* case-insensitive email look up
* Move csv parsing to util.py, add test
* Move verification to util.py, add tests
* Ensure data structure from csv parsing is json-serializable

### CODE REVIEW

**Comment by gvwilson** [comment]
Thanks @sburns - @pbanaszkiewicz can you please review and then merge if you think it's good? This and logins and we're ready to deploy.

**Comment by pbanaszkiewicz** [comment]
Thanks @sburns!

@gvwilson I'll review as soon as learn some [phase portraits](https://en.wikipedia.org/wiki/Phase_portrait), which should be in about hour.

File: workshops/util.py [hunk 1]
```diff
@@ -28,3 +34,75 @@ def earth_distance(pos1, pos2):
...
+    """
+    Read data from CSV and turn it into JSON-serializable list of dictionaries.
+    "Serializability" is required because we put this data into session.  See
+    https://docs.djangoproject.com/en/1.7/topics/http/sessions/ for details.
+    """
+    persons_tasks = []
+    reader = csv.DictReader(uploaded_file)
+    for row in reader:
+        person_fields = dict((col, row[col].strip())
+                             for col in PERSON_UPLOAD_FIELDS)
```
Comment (pbanaszkiewicz) [inline workshops/util.py]: Malformed files (try uploading, for example, a python script) will throw KeyError here: no key 'personal'.

I think the generator may look something like this:

```python
(col, row.get(col, "").strip() for col in PERSON_UPLOAD_FIELDS)
```

But I don't know if this makes any sense… A file without required columns?

Additionally, we need test for this case.

File: workshops/views.py [hunk 2]
```diff
@@ -317,73 +314,6 @@ def person_bulk_add_confirmation(request):
                       context)
```
Comment (pbanaszkiewicz) [inline workshops/views.py]: In lines 288 and 309 (views.py) there's still `_verify_upload_person_task` (should be `verify_upload_person_task`).

Why aren't tests catching this?

File: workshops/util.py [hunk 3]
```diff
@@ -28,3 +34,75 @@ def earth_distance(pos1, pos2):
...
+                Role.objects.get(name=role)
+            except Role.DoesNotExist:
+                errors.append('Role with name {} does not exist.'.format(role))
+            except Role.MultipleObjectsReturned:
+                errors.append('More than one role named {} exists.'
+                              .format(role))
+
+        if email:
+            try:
+                Person.objects.get(email__iexact=email)
```
Comment (pbanaszkiewicz) [inline workshops/util.py]: I'm fine with using case-insensitive match `email__iexact=email`, but I think we might have minor design issue: our DB constraints allow to insert two same emails with different-case…

File: workshops/test/test_util.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+    def test_serializability_of_parsed(self):
+        csv = """personal,middle,family,email\njohn,a,doe,johndoe@email.com\n,jane,a,doe,janedoe@email.com"""
+        person_tasks = self.compute_from_string(csv)
+
+        try:
+            dumps(person_tasks)
+        except Exception:
+            self.fail('Dumping person_tasks to JSON unexpectedly failed!')
+
```
Comment (pbanaszkiewicz) [inline workshops/test/test_util.py]: As I mentioned in https://github.com/swcarpentry/amy/pull/165/files#r23504978, we need to test against the regression with malformed files being uploaded.  It's enough to test for CSV files that have columns with wrong names.

File: workshops/test/test_util.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    def compute_from_string(self, csv_str):
+        ''' wrap up buffering the raw string & parsing '''
+        csv_buf = StringIO(csv_str)
+        # compute & return
+        return upload_person_task_csv(csv_buf)
+
+    def test_basic_parsing(self):
+        ''' See views.PERSON_UPLOAD_FIELDS for field ordering '''
+        csv = """personal,middle,family,email\njohn,a,doe,johndoe@email.com\n,jane,a,doe,janedoe@email.com"""
```
Comment (pbanaszkiewicz) [inline workshops/test/test_util.py]: (Personal preference) I'd rather have this split into multiple lines and not use `"\n"` explicitely. You have `"""` already.

File: workshops/test/test_util.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        # assert
+        self.assertEqual(len(person_tasks), 2)
+
+        person = person_tasks[0]
+        for key in ('person', 'role', 'event', 'errors'):
+            self.assertIn(key, person)
+
+        person_dict = person['person']
+        for key in PERSON_UPLOAD_FIELDS:
+            self.assertIn(key, person_dict)
```
Comment (pbanaszkiewicz) [inline workshops/test/test_util.py]: These two for-loops can be squizzed into:
```python
    self.assertSetEqual(set(('person', 'role', 'event', 'errors')), set(person.keys()))
    self.assertSetEqual(set(PERSON_UPLOAD_FIELDS), set(person_dict.keys()))
```
The advantage is that if `person` or `person_dict` have more keys than we anticipated, we'll get an error.  This ensures that `PERSON_UPLOAD_FIELDS` is up-to-date with `Person` model.

File: workshops/test/test_util.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+        self.assertEqual(person['person']['middle'], '')
+
+    def test_serializability_of_parsed(self):
+        csv = """personal,middle,family,email\njohn,a,doe,johndoe@email.com\n,jane,a,doe,janedoe@email.com"""
+        person_tasks = self.compute_from_string(csv)
+
+        try:
+            dumps(person_tasks)
+        except Exception:
```
Comment (pbanaszkiewicz) [inline workshops/test/test_util.py]: We're using [Django wrapper around dumps](https://docs.djangoproject.com/en/1.7/_modules/django/core/signing/#dumps), not "pure" Python JSON serializer.

Please switch to Django `dumps` and check for proper exception (ie. TypeError).

**Comment by pbanaszkiewicz** [comment]
Hey @sburns,

this PR looks okay. Mostly I need you to fix these small errors and ensure they don't pop out anymore. Tests don't catch them, so we need some tests to test this part of the code.

Other than that there are only some minor improvements to test suite.

Cheers,
Piotr

File: workshops/test/test_util.py (hunk 4 above)
Comment (sburns) [inline workshops/test/test_util.py]: I was looking at `PERSON_UPLOAD_FIELDS` and `static/BulkPersonAddTemplate.csv`...this file don't match the fields as it is. I agree we need to test against malformed csv. But we should generate this template as runtime using this single source of column names.

**Comment by sburns** [comment]
- [ ] switch to django's json dumps
- [ ] squeeze basic parse test into equal sets
- [ ] explicit newline in test csv string
- [ ] generate BulkPersonAddTemplate.csv dynamically using PERSON_UPLOAD_FIELDS as columns.
- [ ] show useful error message for KeyErrors in upload_person_task

File: workshops/test/test_util.py (hunk 4 above)
Comment (pbanaszkiewicz) [inline workshops/test/test_util.py]: I didn't look at `BulkPersonAddTemplate.csv` but it's a great idea to keep it up to date with `PERSON_UPLOAD_FIELDS`.

Maybe we should throw `PERSON_UPLOAD_FIELDS` into `Person` model definition with a comment suggesting to keep that list up-to-date with `Person` fields?

File: workshops/util.py [hunk 8]
```diff
@@ -1,5 +1,11 @@
+import csv
 from math import pi, sin, cos, acos
 
+from .models import Event, Role, Person
+
+PERSON_UPLOAD_FIELDS = ['personal', 'middle', 'family', 'email']
+PERSON_TASK_UPLOAD_FIELDS = PERSON_UPLOAD_FIELDS + ['event', 'role']
```
Comment (pbanaszkiewicz) [inline workshops/util.py]: Please make `PERSON_UPLOAD_FIELDS` and `PERSON_TASK_UPLOAD_FIELDS` immutable by turning them into tuples.

Additionally you can move them into `Person` model definition as stated here: https://github.com/swcarpentry/amy/pull/165#discussion_r23534853

**Comment by sburns** [comment]
@pbanaszkiewicz I think I hit everything, let me know?

**Comment by pbanaszkiewicz** [comment]
@sburns thanks a lot! I'll check this out tomorrow and let you know.

File: workshops/util.py [hunk 9]
```diff
@@ -28,3 +66,84 @@ def earth_distance(pos1, pos2):
...
+            except Event.DoesNotExist:
+                errors.append('Event with slug {} does not exist.'
+                              .format(event))
+        if role:
+            try:
+                Role.objects.get(name=role)
+            except Role.DoesNotExist:
+                errors.append('Role with name {} does not exist.'.format(role))
+            except Role.MultipleObjectsReturned:
+                errors.append('More than one role named {} exists.'
```
Comment (pbanaszkiewicz) [inline workshops/util.py]: Strings in lines 122, 128 and 130 (`errors.append(…)`) should be unicode (ie. `u'Event with slug…'`). Otherwise when uploading non-ASCII characters there would be errors in these lines.

**Comment by pbanaszkiewicz** [comment]
@sburns as part of the code review I found three bugs (including the bug with unicode that I commented in https://github.com/swcarpentry/amy/pull/165#discussion_r23617908).

All three are basically bugs in my original code, and I have fixed them to make sure the code's working, so I'll simply send you a PR so that you can update this PR with our code.

When it comes to your latest additions to this PR, I'm really happy with your code.

To sum up: thank you for your updates and tests. They're great. I'll now send you a PR, so you can merge it and then this PR (#165) should update automatically. Finally we'll be able to successfully merge this PR (#165) into the master branch.

**Comment by pbanaszkiewicz** [comment]
@sburns: here you go https://github.com/sburns/amy/pull/1.
It took longer that I anticipated :/ but there should be less errors now.

**Comment by sburns** [comment]
Merged, thanks @pbanaszkiewicz. @gvwilson I think this is ready to go.

**Comment by pbanaszkiewicz** [comment]
Okay, on the permission given me by @gvwilson in https://github.com/swcarpentry/amy/pull/165#issuecomment-71377130 I shall merge.
